contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,507 @@
+
+Initializers, like external admission hooks, raise the potential for a cluster that cannot
+make progress or heal itself. An initializer on pods could block an emergency fallback
+scheduler from launching a new scheduler pod. An initializer on the endpoints resource could
+prevent masters from registering themselves, blocking an extension API server from observing
+endpoints changes to allow it to watch endpoints. In general, new initializers must be
+careful to not create circular dependencies with the masters.
+
+The ability to set an empty initializer list allows cluster level components to make progress
+in the face of extension. Additional information may need to be returned with an object
+creation error to indicate which component failed to initialize.
+
+
+#### Generic external admission webhook
+
+Existing webhooks demonstrate specific admission callouts for image policy. In general, the `admission.Interface`
+already defines a reasonable pattern for mutations. Admission is much less common for retrieval operations.
+
+Add a new `GenericAdmissionWebhook` that is a list of endpoints that will receive POST operations. The schema
+of the object is modelled after `SubjectAccessReview` and `admission.Interface`. The caller posts the object
+to the server and expects a 200, 204 or 400 response. If the response is 200 or 204, the caller proceeds. If
+the response is 400, the client should interpret the response body to determine the "status" sub field and
+then use that as the response.
+
+```
+POST [arbitrary_url]
+{
+    "kind": "AdmissionReview",
+    "apiVersion": "admission.k8s.io/v1",
+    "spec": {
+        "resource": "pods",
+        "subresource": "",
+        "operation": "Create",
+        "object": {...},
+        "oldObject": nil,
+        "userInfo": {
+            "name": "user1",
+            "uid": "cn=user,dn=ou",
+            "groups": ["g=1"]
+        }
+    },
+    "status": {
+        "status": "Failure",
+        "message": "...",
+        "reason": "Forbidden",
+        "code": 403,
+        ...
+    }
+}
+
+400 Bad Request
+{
+    "kind": "AdmissionReview",
+    "apiVersion": "admission.k8s.io/v1",
+    "spec": {
+        ...
+    },
+    "status": {
+        "status": "Failure",
+        "message": "...",
+        "reason": "Forbidden",
+        "code": 403,
+        ...
+    }
+}
+```
+
+Clients may return 204 as an optimization. `oldObject` may be sent if this is an update operation. This
+API explicitly does not allow mutations, but leaves open the possibility for that to be added in the future.
+
+Each webhook must be non-mutating and will be invoked in parallel. The first failure will terminate processing,
+and the caller may choose to retry calls and so submission must be idempotent.
+
+Because admission is performance critical, the following considerations are taken:
+
+1. The caller may pass the received body bytes AS-IS to the admission controller, so defaulting may not be performed.
```

Comment:
The result of this is likely that the admission controller would do some of its own defaulting, which I'm against. The admission controller would not be aware of default changes and may build in assumptions. I guess it's not the end of the world as long as this remains non-mutating, but I bet it will cause admission controller authors a some heartache.